### PR TITLE
WIP: Update Kubernetes API to support Debug Containers

### PR DESCRIPTION
**Pull request by verb**
**What this PR does / why we need it**: Debug Containers were proposed in kubernetes/community#649 and are to be implemented mostly in the kubelet. This change proposes changes to the Kubernetes API to allow creating a Debug Container via the existing exec subresource.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

**Dependencies:**
- [x] #46261 Debug Containers Feature Gate

### CODE REVIEW

File: pkg/registry/core/pod/strategy.go [hunk 1]
```diff
@@ -425,6 +446,52 @@ func ExecLocation(
...
+		return nil, nil, err
+	}
+	params := url.Values{}
+	if err := streamParams(params, opts); err != nil {
+		return nil, nil, err
+	}
+	loc := &url.URL{
+		Scheme:   nodeInfo.Scheme,
+		Host:     net.JoinHostPort(nodeInfo.Hostname, nodeInfo.Port),
+		Path:     fmt.Sprintf("/podDebug/%s/%s/%s", pod.Namespace, pod.Name, opts.Container),
```
Comment (kargakis) [inline pkg/registry/core/pod/strategy.go]: poddebug?

**Comment by verb** [comment]
One thing I wanted to ask about: the kubelet needs a v1.Container to create the debug container. I think it would be ideal if the client could post a v1.Container to the /debug subresource rather than passing individual fields via http parameters. That's possible to do prior to upgrading the connection, but it doesn't seem well supported by the api machinery and I'm not sure how well that fits in kubernetes API conventions. Advice would be much appreciated.

File: pkg/registry/core/pod/strategy.go (hunk 1 above)
Comment (verb) [inline pkg/registry/core/pod/strategy.go]: most of the other two word kubelet paths use mixed case (e.g. portForward, containerLogs), but then there's "runningpods" so perhaps that's an old convention.

Comment (kargakis) [inline pkg/registry/core/pod/strategy.go]: Yeah, I think we want to stop using camel cased paths

Comment (verb) [inline pkg/registry/core/pod/strategy.go]: ok, thanks. I've made the change and it will appear when I push a rebase

**Comment by verb** [comment]
Update: I'm still working on this. I've reworked the API and will update this PR in the next week or so.

**Comment by verb** [comment]
@lavalamp is unavailable for a bit so I'm going to request a new reviewer.

/unassign @lavalamp

**Comment by derekwaynecarr** [comment]
this work will require us to implement this first, correct?

https://github.com/kubernetes/community/pull/1048

will we implement the above before this pr lands?

**Comment by verb** [comment]
@derekwaynecarr It's not necessary because the features are only loosely coupled. The code I already have uses the existing, kubelet-flagged shared PID support.

We can order them if you'd like, though. I consider Shared PID more urgent, but I'm not sure how long the review of kubernetes/community#1048 will take.

**Comment by verb** [comment]
We're not going to get consensus here any time soon. Will reopen in the future if we figure things out.

**Pull request closed without merging**